### PR TITLE
[SYCL][HIP] Implement mem_advise for HIP

### PR DESCRIPTION
**Pull request by GeorgeWeb**
Adds initial implementation for mem_advise to the HIP adapter.

### CODE REVIEW

**Comment by GeorgeWeb** [comment]
Related intel/llvm PR: https://github.com/intel/llvm/pull/10697

**Comment by JackAKirk** [comment]
I just remembered that this is not consistent with the change to the spec made here: https://github.com/oneapi-src/unified-runtime/pull/854/files

See how the cuda behaviour has been changed here: https://github.com/oneapi-src/unified-runtime/pull/989

I think the behaviour should at least be consistent across adapters.

**Comment by GeorgeWeb** [comment]
> I just remembered that this is not consistent with the change to the spec made here: https://github.com/oneapi-src/unified-runtime/pull/854/files
> 
> See how the cuda behaviour has been changed here: #989
> 
> I think the behaviour should at least be consistent across adapters.

@JackAKirk  Good catch. This PR was moved from intel/llvm where it was open for a while so I kinda missed the UR-spec related changes. Will update, thanks for spotting this.

**Comment by JackAKirk** [comment]
> > I just remembered that this is not consistent with the change to the spec made here: https://github.com/oneapi-src/unified-runtime/pull/854/files
> > See how the cuda behaviour has been changed here: #989
> > I think the behaviour should at least be consistent across adapters.
> 
> @JackAKirk Good catch. This PR was moved from intel/llvm where it was open for a while so I kinda missed the UR-spec related changes. Will update, thanks for spotting this.

no worries, otherwise LGTM.

File: source/adapters/hip/enqueue.cpp [hunk 1]
```diff
@@ -85,6 +85,62 @@ void simpleGuessLocalWorkSize(size_t *ThreadsPerBlock,
...
+  }
+
+  // Handle unmapped memory advice flags
+  if (URAdviceFlags &
+      (UR_USM_ADVICE_FLAG_SET_NON_ATOMIC_MOSTLY |
+       UR_USM_ADVICE_FLAG_CLEAR_NON_ATOMIC_MOSTLY |
+       UR_USM_ADVICE_FLAG_BIAS_CACHED | UR_USM_ADVICE_FLAG_BIAS_UNCACHED)) {
+    return UR_RESULT_ERROR_INVALID_ENUMERATION;
+  }
+
```
Comment (mmoadeli) [inline source/adapters/hip/enqueue.cpp:140]: I would move ` // Handle unmapped memory advice flags` to start of  of the function to report the error earlier.

File: source/adapters/hip/enqueue.cpp [hunk 2]
```diff
@@ -85,6 +85,62 @@ void simpleGuessLocalWorkSize(size_t *ThreadsPerBlock,
     --ThreadsPerBlock[0];
   }
 }
+
+ur_result_t setHipMemAdvise(const void *DevPtr, size_t Size,
+                            ur_usm_advice_flags_t URAdviceFlags,
```
Comment (mmoadeli) [inline source/adapters/hip/enqueue.cpp:90]: can the `Size` be `const`?
Also, in  `urEnqueueUSMAdvise` the argument is `size_t size`, here is  `size_t Size`.

File: source/adapters/hip/enqueue.cpp (hunk 1 above)
Comment (GeorgeWeb) [reply source/adapters/hip/enqueue.cpp:140]: Sure, will do.

**Review by mmoadeli — APPROVED** [review]

**Comment by fabiomestre** [comment]
I have updated the target branch of this PR from the `adapters` branch to the `main` branch.
Development in UR is moving back to `main`. The `adapters` branch will soon be deleted.

**Comment by kbenzie** [comment]
@GeorgeWeb FYI I rebased this on top of main to clean up the history aiming to merge this next.

**Pull request merged**